src/cunumeric/cuda_help.h

```diff
@@ -67,6 +67,20 @@
     check_nccl(result, __FILE__, __LINE__); \
   } while (false)
 
+#ifdef DEBUG_CUNUMERIC
+
+#define CHECK_CUDA_STREAM(stream)              \
+  do {                                         \
+    CHECK_CUDA(cudaStreamSynchronize(stream)); \
+    CHECK_CUDA(cudaPeekAtLastError());         \
```

Comment:
I think this should also be called in the non-debug case, to catch some cases of silent failure, e.g. where a task only does a kernel launch and that launch fails due to `threadsPerBlock > maxThreadsPerBlock` from `__launch_bounds__`. It looks like currently in release mode kernel launches are not guarded at all, and we cannot rely on the Realm context sync to catch the error, because it's not a driver error. This would lead to the error silently causing data corruption.

Having this check in release mode shouldn't be a huge deal, since AFAIU it doesn't block.

We could also just put this runtime check in the Legate task wrapper.

Also, out of curiosity, is there a reason to prefer `peekAtLastError` vs `getLastError`?

Reply:
I guess the expected use for `CHECK_CUDA_STREAM` is not to call it once at the end of the task, but instead call it after every kernel launch, so we can pinpoint the error faster in debug mode. In that case the definition makes sense. I still think we should put one use of `peekAt/getLastError` in the task postamble, that runs even in release mode.

Reply:
> I still think we should put one use of peekAt/getLastError in the task postamble, that runs even in release mode.

That's a good suggestion, but I'd probably turn it on optionally, as we will eventually get rid of the synchronization that we're currently inserting at the end of every GPU task.

Reply:
I think the usecases I am considering for this exit check are independent of whether we synchronize at the end of the task.

I assume that, no matter what method we use to encapsulate a task's CUDA effects within the envelope of realm-managed effects (e.g. using the realm CUDART hijack, or doing an explicit context sync), a driver-level error check will always happen before task completion. AFAIU, errors like "too many threads per block" signal immediately at kernel launch, and don't even enter the driver, thus are not caught by the existing exit checks. The extra exit check I am suggesting would be a lightweight way to catch such runtime-only errors, so they don't cause silent failures.

Since the `peekAt/getLastError` check will not cause any extra blocking, and it would help catch error conditions that currently go unnoticed, I would prefer to always include it.

Reply:
All this said, this PR doesn't need to be blocked on this issue. I will start an issue about this on legate.core and move this discussion there.

Reply:
> I assume that, no matter what method we use to encapsulate a task's CUDA effects within the envelope of realm-managed effects (e.g. using the realm CUDART hijack, or doing an explicit context sync), a driver-level error check will always happen before task completion. AFAIU, errors like "too many threads per block" signal immediately at kernel launch, and don't even enter the driver, thus are not caught by the existing exit checks. The extra exit check I am suggesting would be a lightweight way to catch such runtime-only errors, so they don't cause silent failures.
Agreed. I forgot to mention that the added macro also has a stream synchronize call that we don't want to introduce in release mode, and that's why I mentioned about the choice of post-task synchronization. I can and will certainly add a similar facility to Legate core, but without the device synchronize as `CUDA_CHECK_STREAM` that I currently have.